Improved docs for `torch.linalg`

lezcano opened the pull request:
This PR tries to make the docs of `torch.linalg` have/be:
- More uniform notation and structure for every function.
- More uniform use of back-quotes and the `:attr:` directive
- More readable for a non-specialised audience through explanations of the form that factorisations take and when would it be beneficial to use what arguments in some solvers.
- More connected among the different functions through the use of  the `.. seealso::` directive.
- More information on when do gradients explode / when is a function silently returning a wrong result / when things do not work in general

I tried to follow the structure of "one short description and then the rest" to be able to format the docs like those of `torch.` or `torch.nn`. I did not do that yet, as I am waiting for the green light on this idea:
https://github.com/pytorch/pytorch/issues/54878#issuecomment-816636171

What this PR does not do:
- Clean the documentation of other functions that are not in the `linalg` module (although I started doing this for `torch.svd`, but then I realised that this PR would touch way too many functions).

Fixes https://github.com/pytorch/pytorch/issues/54878

cc @mruberry @IvanYashchuk

lezcano committed:
Spectral decompositions

lezcano committed:
First draft

lezcano committed:
Trailing spaces

lezcano committed:
Break lines

lezcano committed:
split line

lezcano committed:
Trailing spaces

xwang233 reviewed (commented):
Thanks for the PR! I left some minor comments for the cusolver backend of `cholesky` (`potrf`). Also, `cholesky_solve` and `cholesky_inverse` both use corresponding cusolver backend functions for batch_size 1. 

I will check other details later. Thanks!

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8292,25 +8292,30 @@ def merge_dicts(*dicts):
+matrices :attr:`input`.
+
+The singular value decomposition is represented as a
+namedtuple `(U, S, V)`, such that `input = U diag(S) Vᴴ`.
```
I'm curious what your thoughts are on https://github.com/pytorch/pytorch/issues/54878#issuecomment-821962806, internally we're hoping to get rid of unicode like Vᴴ because we're concerned about supporting it on multiple platforms.

Lezcano replied:
I think LaTeX is definitely the way to go. These `torch.svd` docs are not as polished as  `torch.linalg.svd` (I should have updated them to look the same).
In the end, I just used unicode in some places to write inline maths. The unicode characters that I used are the following:
`≤ ≥ σ₁ σᵢ σₙ S⁻¹ · ε`.

I have to say that inline maths looks nicer in unicode, but I used it so little that it'd be no problem to change it by code or LaTeX. I'll update the PR not to use unicode.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8292,25 +8292,30 @@ def merge_dicts(*dicts):
 where `Vᴴ` is the transpose of `V` for real inputs,
 and the conjugate transpose of `V` for complex inputs.
-If :attr:`input` is a batch of matrices, then `U`, `S`, and `V` are also
+If :attr:`input` is a batch of matrices, then `U`, `S`, `V` are also
```
why remove the "and" here?

Lezcano replied:
I just figured that it reads better `U`, `S`, `V`, even tough it is less grammatically correct. Should I revert?

mruberry replied:
Not a big deal either way; I would stick with the grammatically correct version myself

Lezcano replied:
Reverted

Lezcano replied:
And made uniform throughout the docs

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8292,25 +8292,30 @@ def merge_dicts(*dicts):
 batched with the same batch dimensions as :attr:`input`.
 
+The singular values are returned in descending order. If :attr:`input` is a batch of matrices,
+the singular values of each matrix in the batch are returned in descending order.
```
", the" -> "then the"

"are returned" -> "are also returned"

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8292,25 +8292,30 @@ def merge_dicts(*dicts):
-Supports :attr:`input` of float, double, cfloat and cdouble data types.
-The dtypes of `U` and `V` are the same as :attr:`input`'s. `S` will
+Supports :attr:`input` of float, double, cfloat and cdouble dtypes.
+The dtype of `U` and `V` is the same as that of :attr:`input`. `S` will
```
"of \`U\` and \`V\` is" -> "of the output \`U\` and \`V\` tensors will be"

"\`S\` will" -> "The output tensor \`S\` will"

Lezcano replied:
Corrected and made uniform through all the functions that use this construction

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8329,38 +8334,34 @@ def merge_dicts(*dicts):
 
-.. note:: The implementation of :func:`torch.linalg.svd` on CPU uses LAPACK's routine `?gesdd`
-          (a divide-and-conquer algorithm) instead of `?gesvd` for speed. Analogously,
+.. note:: The implementation of :func:`torch.linalg.svd` on CPU uses LAPACK's routine `gesdd`
```
"routine" -> ""

Lezcano replied:
Corrected in all the docs

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8329,38 +8334,34 @@ def merge_dicts(*dicts):
-          (a divide-and-conquer algorithm) instead of `?gesvd` for speed. Analogously,
+.. note:: The implementation of :func:`torch.linalg.svd` on CPU uses LAPACK's routine `gesdd`
+          (a divide-and-conquer algorithm) instead of `gesvd` for speed. Analogously,
           on GPU, it uses cuSOLVER's routines `gesvdj` and `gesvdjBatched` on CUDA 10.1.243
```
"routines" -> ""

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8329,38 +8334,34 @@ def merge_dicts(*dicts):
-.. note:: The implementation of :func:`torch.linalg.svd` on CPU uses LAPACK's routine `?gesdd`
-          (a divide-and-conquer algorithm) instead of `?gesvd` for speed. Analogously,
+.. note:: The implementation of :func:`torch.linalg.svd` on CPU uses LAPACK's routine `gesdd`
+          (a divide-and-conquer algorithm) instead of `gesvd` for speed. Analogously,
```
How is this analogous?

Lezcano replied:
It isn't.  Corrected.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8329,38 +8334,34 @@ def merge_dicts(*dicts):
-          (a divide-and-conquer algorithm) instead of `?gesvd` for speed. Analogously,
+.. note:: The implementation of :func:`torch.linalg.svd` on CPU uses LAPACK's routine `gesdd`
+          (a divide-and-conquer algorithm) instead of `gesvd` for speed. Analogously,
           on GPU, it uses cuSOLVER's routines `gesvdj` and `gesvdjBatched` on CUDA 10.1.243
```
"on GPU, it" -> "The CUDA implementation uses"

Lezcano replied:
Done. Should I replace `GPU` by `CUDA` throughout the docs?

mruberry replied:
Being more consistent about CUDA would be great, yes.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8329,38 +8334,34 @@ def merge_dicts(*dicts):
 
-.. note:: The implementation of :func:`torch.linalg.svd` on CPU uses LAPACK's routine `?gesdd`
-          (a divide-and-conquer algorithm) instead of `?gesvd` for speed. Analogously,
+.. note:: The implementation of :func:`torch.linalg.svd` on CPU uses LAPACK's routine `gesdd`
```
"The implementation" -> "The CPU implementation"

"on CPU" -> ""

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8329,38 +8334,34 @@ def merge_dicts(*dicts):
-
-.. warning:: For complex-valued :attr:`input` the singular value decomposition is not unique,
-             as `U` and `V` may be multiplied by an arbitrary phase factor :math:`e^{i \phi}` on every column.
+.. warning:: The singular values of a matrix are not unique. Any pair of left and right singular values may
```
Should this be a warning or a note? Warning seems kind of extreme?

mruberry replied:
This is interesting. 

We could, for example, return the singular values in descending order, and then the singular values would be unique, right? It's just that the columns of U (the left-singular vectors) and the columns of V (the right-singular vectors) are not unique. Is that what's meant by "left singular values" and "right singular values" above? I'm not familiar with that language.

Lezcano replied:
Yes. The singular values are always unique (same with eigenvalues for `eig` /  `eigh`). The singular vector / eigenvectors are what causes problems. I had corrected this errata in `torch.linalg.svd`, but I had forgotten to port it to `torch.svd`.

This now reads like this, tell me what you think:
![image](https://user-images.githubusercontent.com/3291265/115226755-826da980-a107-11eb-8b65-4089b68364c9.png)

About the warning vs note, I think that it's fine to leave it as a warning. It is surprisingly tricky to correctly use `U` and `Vh` in the loss function without them biting you due to the lack of unicity reasons. People expect functions to be continuous (and even differentiable) and `U` and `Vh` are not differentiable nor continuous, so I'd suggest to leave it as a warning.

mruberry replied:
The note at the comment is excellent (I commented there).

OK, leaving it as a warning sounds OK.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8329,38 +8334,34 @@ def merge_dicts(*dicts):
-.. warning:: For complex-valued :attr:`input` the singular value decomposition is not unique,
-             as `U` and `V` may be multiplied by an arbitrary phase factor :math:`e^{i \phi}` on every column.
+.. warning:: The singular values of a matrix are not unique. Any pair of left and right singular values may
+             be multiplied by `-1` in the real case or by a factor of norm `1` :math:`e^{i \phi}` in the complex case.
```
How would these elements be paired? Is it corresponding columns in the U and V matrices? 

"in the real case" -> "when U and V are real-valued" 
"in the complex case" -> "when U and V are complex"

I'm not sure I understand the "factor of norm 1 e^(i * phi)" scaling?

Lezcano replied:
What about making this more explicit:
![image](https://user-images.githubusercontent.com/3291265/115227618-98c83500-a108-11eb-94ff-85b8310036f6.png)

Compare this with the proposal from https://github.com/pytorch/pytorch/pull/56265#discussion_r615752325

mruberry replied:
This is much clearer and really excellent. I would consider making it a note, however, since "warning" should probably be reserved for when the user is expected to take an action.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8329,38 +8334,34 @@ def merge_dicts(*dicts):
           and later, and MAGMA's routine `gesdd` on earlier versions of CUDA.
 
 .. note:: The returned `U` will not be contiguous. The matrix (or batch of matrices) will
           be represented as a column-major matrix (i.e. Fortran-contiguous).
```
"The matrix or batch of matrices will be column-major (i.e. Fortran-contiguous).

Lezcano replied:
I figured it'd be best to remove all this comments about the output matrix not being contiguous. I do not think they will be useful to almost anyone, and they are more of an implementation detail. If we tell which LAPACK / CuSOLVER / MAGMA routines we are using, I'm sure that the user that wants to squeeze the most out of these functions will go and check the documentation for those. What do you think?

mruberry replied:
Sounds like a reasonable solution

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8329,38 +8334,34 @@ def merge_dicts(*dicts):
-             as `U` and `V` may be multiplied by an arbitrary phase factor :math:`e^{i \phi}` on every column.
+.. warning:: The singular values of a matrix are not unique. Any pair of left and right singular values may
+             be multiplied by `-1` in the real case or by a factor of norm `1` :math:`e^{i \phi}` in the complex case.
              The same happens when :attr:`input` has repeated singular values, where one may multiply
```
I'm not sure how this paragraph connects to the previous paragraph. The previous paragraph states a mathematical property of the SVD. But then this paragraph begins with "The same happens when..." But what is "the same"?

Lezcano replied:
It comes to say that when the matrix has repeated singular values, there is a second source of "non uniqueness" for the singular vectors.

What's really happening here geometricaly is that SVD (and  `eigh`) work with subspaces, but we choose orthogonal bases to represent them.

If you have a linear subspace, if it's 1D (a line), it can be represented via an orthogonal basis choosing the vector of norm 1 in the direction of the subspace (the direction of the line). Of course, you have two choices, "point to the right" and "point to the left", that is, choosing `v` or choosing `-1*v = -v`.
If it has more dimensions, say it has k dimensions, you can choose a basis, v_1, ... , v_k, but you can rearrange it as v_2, v_1, v_3, ... , v_k, and it will still span the same subspace. Even more, you can rotate the vectors inside the subspace and the new set of rotated vectors will also be orthogonal and span the same subspace. This is this non-uniqueness.

In the complex case this is even worse, as a 1D complex subspace has two real dimension, so you have the problem with being able to "rotate" the basis even in 1D (that's what the multiplication by e^{i \phi}  is really doing).

Lezcano replied:
I tried to address this in a more explicit way in https://github.com/pytorch/pytorch/pull/56265#discussion_r615756832

mruberry replied:
Combining these thoughts into the warning above makes them much clearer and logically connects the docs

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8329,38 +8334,34 @@ def merge_dicts(*dicts):
-             may produce different `U` and `V` tensors.
+             may produce different `U` and `Vh` tensors.
+
+.. warning:: The gradients with respect to `U` and `Vh` will only be finite when :attr:`input`
```
"The gradients with respect to" -> "Gradients computed using"

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8329,38 +8334,34 @@ def merge_dicts(*dicts):
+
+.. warning:: The gradients with respect to `U` and `Vh` will only be finite when :attr:`input`
+             does not have zero or repeated singular values.
+             If the distance between any two singular values is close to zero, the gradients with respect to
```
"Further, if"

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8329,38 +8334,34 @@ def merge_dicts(*dicts):
+.. warning:: The gradients with respect to `U` and `Vh` will only be finite when :attr:`input`
+             does not have zero or repeated singular values.
+             If the distance between any two singular values is close to zero, the gradients with respect to
+             `U` and `V` will be numerically unstable, as they depends on
```
"depends" -> "depend"

Lezcano replied:
I have changed it to that, but I think that is unnecessarily verbose. I would prefer the shorter version for conciseness.

mruberry replied:
OK, just be sure to change "depends" to "depend"

Lezcano replied:
I will update the docs of `torch.svd` to match those of `torch.linalg.svd` once we make them stable

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8329,38 +8334,34 @@ def merge_dicts(*dicts):
+             does not have zero or repeated singular values.
+             If the distance between any two singular values is close to zero, the gradients with respect to
+             `U` and `V` will be numerically unstable, as they depends on
+             :math:`\frac{1}{\min_{i \neq j} \sigma_i^2 - \sigma_j^2}`. The same happens when :attr:`input`
```
What is "the same" here?

Lezcano replied:
Corrected. Tell me what you think

mruberry replied:
I'm not sure I see the correction?

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8329,38 +8334,34 @@ def merge_dicts(*dicts):
+             does not have zero or repeated singular values.
+             If the distance between any two singular values is close to zero, the gradients with respect to
+             `U` and `V` will be numerically unstable, as they depends on
+             :math:`\frac{1}{\min_{i \neq j} \sigma_i^2 - \sigma_j^2}`. The same happens when :attr:`input`
```
Are the terms in this mathematical expression defined?

Lezcano replied:
Corrected. Tell me what you think.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8329,38 +8334,34 @@ def merge_dicts(*dicts):
+             If the distance between any two singular values is close to zero, the gradients with respect to
+             `U` and `V` will be numerically unstable, as they depends on
+             :math:`\frac{1}{\min_{i \neq j} \sigma_i^2 - \sigma_j^2}`. The same happens when :attr:`input`
+             has small singular values, as these gradients also depend on `S⁻¹`.
```
What is `S^{-1}`?

Lezcano replied:
Changed to `
> The gradient will also be numerically unstable when :attr:`input` has small singular values, as it also depends on :math:`\frac{1}{\sigma_i}`.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8292,25 +8292,30 @@ def merge_dicts(*dicts):
-`m` and `n`, then the returned `U` and `V` matrices will contain only
-`min(n, m)` orthonormal columns.
+value decomposition. In this case, if :attr:`input` is of size `(*, m, n)`,
+the returned `U`, `V` will be of size `(*, m, min(m, n))` and `(*, n, min(m, n))`
```
"then \`U\` will have shape ..., and \`V\` will have shape..."

Lezcano replied:
Applied the rule "prefer shape over size" to the documentation.

Lezcano replied:
Also prefer "has shape XX" to "is of shape XX"

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8292,25 +8292,30 @@ def merge_dicts(*dicts):
-value decomposition. In this case, if the last two dimensions of :attr:`input` are
-`m` and `n`, then the returned `U` and `V` matrices will contain only
-`min(n, m)` orthonormal columns.
+value decomposition. In this case, if :attr:`input` is of size `(*, m, n)`,
```
"is of size" -> "has shape"

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -12,55 +12,63 @@
-Computes the Cholesky decomposition of a Hermitian (or symmetric for real-valued matrices)
-positive-definite matrix or the Cholesky decompositions for a batch of such matrices.
-Each decomposition has the form:
+Computes the Cholesky decomposition of a complex Hermitian (or real symmetric) positive-definite matrix or batch of matrices.
```
"or batch of matrices" -> "or batch of such matrices"

Lezcano replied:
Unified to:

No restrictions:
> Computes the XX of a matrix or of a batch of matrices 

Simple restrictions:
> Computes the XX of a square matrix or of a batch of square matrices.

Complex restrictions:
> Computes the XX of a YY matrix or of a batch of such matrices.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -12,55 +12,63 @@
-Computes the Cholesky decomposition of a Hermitian (or symmetric for real-valued matrices)
-positive-definite matrix or the Cholesky decompositions for a batch of such matrices.
-Each decomposition has the form:
+Computes the Cholesky decomposition of a complex Hermitian (or real symmetric) positive-definite matrix or batch of matrices.
```
"(or real symmetric)" -> "or real symmetric"

Lezcano replied:
Corrected throughout the docs

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -12,55 +12,63 @@
+    A = LL^{\text{H}}
+
+where :math:`L` is a lower triangular matrix and
+:math:`L^{\text{H}}` denotes the conjugate transpose in the complex case and the transpose in the real case.
```
"denotes" -> "is"

"the conjugate transpose of L when L is complex, and the transpose of L when L is real-valued."

Lezcano replied:
Corrected throughout the docs.

Lezcano replied:
After seeing the final version, I changed it for:
>" the conjugate transpose when L is complex, and the transpose when L is real-valued."

I think it is more readable as it avoids having to mention the matrix 5 times in the same sentence.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -12,55 +12,63 @@
-          If :attr:`input` is a batch of matrices, then the error message will include the batch index
-          of the first matrix that is not Hermitian positive-definite.
+        :func:`torch.linalg.eigh` for a different decomposition of a Hermitian matrix.
+        The eigenvalue decomposition gives more information about about the matrix but it
```
"about about" -> "about"

"but it is slower" -> "but is slower"

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -80,96 +88,97 @@
 
-Computes the multiplicative inverse matrix of a square matrix :attr:`input`, or of each square matrix in a
-batched :attr:`input`. The result satisfies the relation:
+Computes the `inverse`_ of square matrix or batch of matrices
```
"of square matrix or batch of matrices" -> "of a square matrix or of a batch of square matrices."

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -80,96 +88,97 @@
-.. note:: The inverse matrix is computed using LAPACK's `getrf` and `getri` routines for CPU inputs. For CUDA
-          inputs, cuSOLVER's `getrf` and `getrs` routines as well as cuBLAS' `getrf` and `getri` routines are
-          used if CUDA version >= 10.1.243, otherwise MAGMA's `getrf` and `getri` routines are used instead.
+        :func:`torch.linalg.pinv` computes the Moore-Penrose pseudoinverse of a general matrix
```
period needed at end of this sentence.

what is a "general matrix"?

Lezcano replied:
Added periods in all the `seealso` that I missed.

I changed "general" for "of a (not necessarily square) matrix".

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -80,96 +88,97 @@
 
-.. note:: If :attr:`input` is a non-invertible matrix or non-square matrix, or batch with at least one such matrix,
-          then a RuntimeError will be thrown.
+        :func:`torch.linalg.solve` computes the action of the inverse on another matrix.
```
What does "the action of the inverse on another matrix" mean?

Lezcano replied:
That term is indeed too mathematical. Changed to:

>:func:`torch.linalg.lstsq` computes :attr:`input`\ `.inv() @ \ `:attr:`other`.
Using :func:`torch.linalg.lstsq` is faster and more stable than computing the
inverse and then multiplying.

Updated accordingly in `linalg.inv()` and `linalg.tensorinv()`.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -80,96 +88,97 @@
-.. note:: If :attr:`input` is a non-invertible matrix or non-square matrix, or batch with at least one such matrix,
-          then a RuntimeError will be thrown.
+        :func:`torch.linalg.solve` computes the action of the inverse on another matrix.
+        This is faster and more stable than computing the inverse and then multiplying
```
Period need at end of this sentence

mruberry replied:
I'm not in love with this sentence. Maybe it can be combined with the previous sentence?

Lezcano replied:
I rewrote it going more to the point as:
> :func:torch.linalg.solve computes ``input.inv() @ other``.
It is always prefered to use :func:`torch.linalg.solve` when possible, as it is
faster and more stable than computing the inverse and then multiplying.

Same for `pinv` and `tensorinv`

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -80,96 +88,97 @@
-    input (Tensor): the square `(n, n)` matrix or the batch of such matrices of size
-                    `(*, n, n)` where `*` is one or more batch dimensions.
+    input (Tensor): tensor of size `(*, n, n)` where `*` is zero or more batch dimensions
+                    consisting of `(n, n)` matrices.
```
What does the "consisting of..." portion of this explain? Isn't it explicit in the shape specification?

Lezcano replied:
I agree, I'll delete it.

I think that this went in when doing a big edit of the docs for the `input` tensor. In some other cases it makes sense to day "Tensor of shape XX where `*` is blababla consisting of Hermitian or symmetric matrices", that is, when there are extra conditions on each individual matrix.

I'll correct it in all the docs.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -80,96 +88,97 @@
+    out (Tensor, optional): output tensor. Ignored if `None`. Default: `None`.
+
+Raises:
+    RuntimeError: if the :attr:`input` matrix or any matrix in the batch :attr:`input` is not invertible
```
Period needed

mruberry replied:
"if the input matrix is not invertible or any matrix in a batched input is not invertible"

Lezcano replied:
Applied to all the `RuntimeError`s

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -80,96 +88,97 @@
-        dtype=torch.complex128)
-    >>> torch.max(torch.abs(z - torch.eye(4, dtype=torch.cdouble))) # Max non-zero
+    >>> z = x @ y
+    >>> torch.dist(z, torch.eye(4, dtype=torch.cdouble))
```
This is a really nice update

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -80,96 +88,97 @@
 
-Computes the determinant of a square matrix :attr:`input`, or of each square matrix
-in a batched :attr:`input`.
+Computes the determinant of a matrix or batch of matrices.
```
The matrices have to be square, right?

mruberry replied:
"determinant of a matrix or the determinants of a batch of matrices."

mruberry replied:
Maybe the template for batched support should be:

"Computes the determinant of square matrices. Supports batched and unbatched :attr:\`input\`."

?

Lezcano replied:
I really like this template!

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8292,25 +8292,30 @@ def merge_dicts(*dicts):
 svd(input, some=True, compute_uv=True, *, out=None) -> (Tensor, Tensor, Tensor)
 
 Computes the singular value decomposition of either a matrix or batch of
```
"of a matrix or the decompositions of a batch of matrices input."

Otherwise this sentence would say there's one SVD for the batch

Lezcano replied:
I'll apply the template you suggested here https://github.com/pytorch/pytorch/pull/56265#discussion_r615376640

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -80,96 +88,97 @@
+.. seealso::
+
+        :func:`torch.linalg.slogdet` computes the sign and logarithm of the norm of
+        the determinant of a square matrix
```
period needed

if slogdet supports batching maybe language like:

"computes the signs and logarithms of the norms of the determinants of square matrices" 

would be more accurate?

Lezcano replied:
Agree. Will update all the `seealso`s

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -204,31 +213,42 @@
-Calculates the sign and natural logarithm of the absolute value of a square matrix's determinant,
-or of the absolute values of the determinants of a batch of square matrices :attr:`input`.
-The determinant can be computed with ``sign * exp(logabsdet)``.
+Computes the sign and logarithm of the norm of the determinant of a square matrix or batch of square matrices.
```
Let's try out the above template here to see if we like it more (I'm eager to hear your thoughts on templating):

"Computes the sign and logarithm of the norm of the determinant of square matrices. Supports batched and unbatched :attr:\`input\`.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -204,31 +213,42 @@
+Computes the sign and logarithm of the norm of the determinant of a square matrix or batch of square matrices.
 
-Supports input of float, double, cfloat and cdouble datatypes.
+For a complex :attr:`input`, as the determinant is complex in this case, it returns the angle
```
This is confusing. Is this disagreeing with the first sentence or elaborating on a mathematical perspective? If providing a mathematical perspective then the paragraph should explain it's providing a perspective and possibly be put into a note.

Lezcano replied:
It sort of "disagrees". This function is tricky, as for real inputs it's natural to talk about the "sign of the determinant". For complex inputs, that concept of sign turns into the concept of angle. 
It makes sense to put the two cases separately, even though "sign" the real case is just the complex case on real matrices. Now, if we put the two on the same line, it'd be too long of a "short description". I am not sure how to marry all these things really.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -204,31 +213,42 @@
+and the logarithm of the modulus of the determinant.
 
-.. note:: When given inputs on a CUDA device, this function synchronizes that device with the CPU.
+The returend value decomposition is represented as a namedtuple `(sign, logabsdet)`.
```
"returend" -> "returned"

mruberry replied:
"is represented as a namedtuple" -> "is a named tuple"

Lezcano replied:
This is used throughout the docs. I used "namedtuple" because it is what was used before. Should I change it to "named tuple"?

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -204,31 +213,42 @@
-.. note:: When given inputs on a CUDA device, this function synchronizes that device with the CPU.
+The returend value decomposition is represented as a namedtuple `(sign, logabsdet)`.
+If :attr:`input` is a batch of matrices, then `sign`, `logabsdet` are also batched with the same
+batch dimensions. The determinant can be recovered as `sign * exp(logabsdet)`.
```
Is this sentence about how the "determinant can be recovered" so logically connect to batching? It seems separate. 

Further, what does it mean to "recover" the determinant in this context? Isn't this just a note about how the determinant or determinants can be computed using the results of this operation? That seems like a mathematical note.

Lezcano replied:
Completely agree. I'm moving this to a note.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -204,31 +213,42 @@
 
-.. note:: The determinant is computed using LU factorization. LAPACK's `getrf` is used for CPU inputs,
-          and MAGMA's `getrf` is used for CUDA inputs.
+Supports :attr:`input` of float, double, cfloat and cdouble dtypes.
```
The template I've suggested above might want to merge its support statement into this one:

"Supports batched and unbatched :attr:\`input\` with dtype float, double, cfloat or cdouble."

?

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -204,31 +213,42 @@
-.. note:: The determinant is computed using LU factorization. LAPACK's `getrf` is used for CPU inputs,
-          and MAGMA's `getrf` is used for CUDA inputs.
+Supports :attr:`input` of float, double, cfloat and cdouble dtypes.
+`logabsdet` will always be real-valued, even if :attr:`input` is complex.
```
This seems like it should be connected with the prior discussion of sign and logabsdet. Logically this function's documentation jumps around a lot.

I would expect notes to generally go after regular documentation, and the "supports" statement to go after the opening definition + mathematical statement (if any). Then this discussion of the return values could precede the notes, maybe?

Lezcano replied:
In all the functions I tried to follow the following schema (open to discussion about whether it's reasonable or there are better ones):
- One line summary
- Mathematical statement of what it does. In LaTeX if it's non-trivial (e.g. for factorisations). Here both the real and complex cases are treated.
- Describe the return if it's a tuple. Put the return size if it is non-trivial (e.g. describe it for `qr` but not for `norm` or `inv`). 
- Describe the arguments and what they do. Sometimes the size of the return is described here if the size of the return depends on some arguments
- Describe the supported dtypes and the dtype of the outputs if it's not the same as that of the inputs (as in this case).

It could be argued that that last step could be moved up. I would not mind to do so.

I have avoided putting too much information in the `Arguments` section, and I have prefered to completely avoid using the `Returns` section. The reasoning behind this is that, in my experience, they are too far back in the documentation. Furthermore, when you need to describe some behaviour that is particularly tricky (e.g. the norms in `linalg.norm` / `linalg.cond`) the indentation makes everything way too crammed.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -204,31 +213,42 @@
+
+.. note:: This function is computed using :func:`torch.lu`.
+
+.. note:: For matrices that have zero determinant, this returns `(0, -inf)`.
```
"have zero determinant" -> "a determinant of zero"

mruberry replied:
I think this unbatched vs. batched templating can be simplified, and this note reduced to:

"When a matrix has a determinant of zero this computes the sign of its determinant as zero and the natural logarithm of the norm of its determinant as negative infinity."

Lezcano replied:
What about the more concise:
>  When a matrix has a determinant of zero, it returns `(0., -inf)`.
?

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -204,31 +213,42 @@
-          the zero determinant have sign 0 and the natural logarithm of the absolute value of the determinant -inf.
+.. note:: When :attr:`input` is on a CUDA device, this function synchronizes that device with the CPU.
+
+.. note:: This function is computed using :func:`torch.lu`.
```
This note can probably be combined with the previous note. In the future we'll have an "Implementation details" section of the docs, so everything we'd put into that section can go into the same note.

Lezcano replied:
This is a great point! Shorter is definitely better.

I have merged the implementation with the synchronisation in all the docs. The "implementation details" section would be a great one, certainly.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -204,31 +213,42 @@
+
+.. seealso::
+
+        :func:`torch.linalg.det` computes the determinant of a square matrix
```
"computes the determinant of square matrices." ?

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -204,31 +213,42 @@
-    input (Tensor): the input matrix of size :math:`(n, n)` or the batch of matrices of size :math:`(*, n, n)`
-                    where :math:`*` is one or more batch dimensions.
+    input (Tensor): tensor of size `(*, n, n)` where `*` is zero or more batch dimensions
+                    consisting of `(n, n)` matrices.
```
the "consisting" here seems redundant with the shape information

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -204,31 +213,42 @@
+    out (tuple, optional): output tuple of two tensors. Ignored if `None`. Default: `None`.
 
 Returns:
     A namedtuple (sign, logabsdet) containing the sign of the determinant and the natural logarithm
```
"containing the sign and natural logarithm of the norm of an input matrix's determinant or the signs and natural logarithms of the norms of input matrices' determinants."

Lezcano replied:
I'm deleting this return for the reasons given here: https://github.com/pytorch/pytorch/pull/56265#discussion_r615922987

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -248,42 +268,60 @@
 
-Computes the eigenvalues and eigenvectors of a square
-matrix :attr:`input`, or of each such matrix in a batched :attr:`input`.
+Computes the eigenvalue decomposition of a square matrix or batch of square matrices.
```
Continuing to experiment with templating, this could be "Computes the eigenvalue decomposition of square matrices."

mruberry replied:
Another option would be to talk exclusively about the batched case and treat single "unbatched" matrices as just having empty batch dimensions. That might be a lot simpler and less conceptually confusing to both writers and readers?

Lezcano replied:
In my opinion, it's the other way around. It's much easier to talk about the non-batched case, as the examples you write them for one matrix such as `A = VSV^{-1}`. Writing this for batched inputs is unnecessarily verbose.

I think that we should find patterns that concentrate in conveying the idea of what happens for one matrix. In particular, the we should balance
1) Making clear that everything works for batches
2) Making clear that batches are "just an optimisation", "something convenient", and that what's important to understand is the matrix case.

I really liked the template you proposed above. As such, I'd write this one as:

> Computes the eigenvalues and eigenvectors of a square matrix if they exist. Supports batched and unbatched inputs.

Note how I also slipped in the "if they exist", as the eigenvalues as defined here only exist for diagonalisable matrices. This point is addressed below, but I think it's good to point out that some operations may not be well-defined (e.g. the inverse of a non-invertible matrix). I elaborate on this point in the documentation of this operation.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -248,42 +268,60 @@
+
+    A = V \operatorname{diag}(L) V^{-1}
+
+This decomposition exists if :math:`A` is diagonalizable (for example, when all its eigenvalues are different).
```
what does "diagonalizable" mean?

mruberry replied:
what happens if the decomposition doesn't exist?

Lezcano replied:
That's a tricky one, as it is defined as "matrices for which this function would succeed" :D That's why I gave there a simple condition that works in almost any practical situation: A matrix with different eigenvalues is diagonalisable.

Would it be fine to link to the following wikipedia article?
https://en.wikipedia.org/wiki/Diagonalizable_matrix#Definition

If the decomposition does not exist, then the eigenvalues returned by this function are incorrect. The point to be made here is that this is not a problem of the algorithm, but a problem intrinsical to the matrix.

Lezcano replied:
I have updated this point and the warning below. How does it read now?

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -248,42 +268,60 @@
 
-The eigenvalues are not necessarily sorted.
-For real-valued input the resulting eigenvalues and eigenvectors will always be of complex type.
+The returned decomposition is represented as a namedtuple `(eigenvalues, eigenvectors)`,
```
"is represented as a namedtuple" -> "is a named tuple"

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -248,42 +268,60 @@
-The eigenvalues are not necessarily sorted.
-For real-valued input the resulting eigenvalues and eigenvectors will always be of complex type.
+The returned decomposition is represented as a namedtuple `(eigenvalues, eigenvectors)`,
+which corresponds to :math:`L`, :math:`V` above. If :attr:`input` is a batch of matrices,
```
what corresponds to what?

Lezcano replied:
L = eigenvalues
V = eigenvectors

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -248,42 +268,60 @@
 
-Supports input of float, double, cfloat and cdouble dtypes.
+Supports :attr:`input` of float, double, cfloat and cdouble dtypes.
+The eigenvalues and eigenvectors will always be complex-valued, even if :attr:`input` is real.
```
This phrasing makes this sentence appear to be a mathematical statement about eigenvalues and eigenvectors, but I think this is trying to make a statement about the output tensors, so it needs to explicitly relate these ambiguous names to the tensors.

"The output \`eigenvalues\` and \`eigenvectors\` tensors will always be complex, even if :attr:\`input\` is real-valued, since the eigenvalues and eigenvectors of real-valued matrices may be complex."

mruberry replied:
This may also be an opportunity to extend that sentence to explain that PyTorch's dtype behavior is divergent from NumPy's. That behavior might also be so nuanced that it could go in a note.

Lezcano replied:
Following a previous suggestion, I edited this as:

> The output tensors `eigenvalues` and `eigenvectors` will always be complex-valued, even when :attr:      `input` is real.

I'll add a note explaining this behaviour, as it's done for Hermitian matrices in `eigh`, where the matrices are always real even if the matrix is complex.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -248,42 +268,60 @@
+.. note:: When :attr:`input` is on a CUDA device, this function synchronizes that device with the CPU.
 
-.. note:: The eigenvalues and eigenvectors are computed using LAPACK's and MAGMA's `geev` routine.
+.. note:: This function is computed using LAPACK's and MAGMA's `geev` routine for
```
"routine" -> ""

If we want to consistently specify what these names refer to then we should consistently say "function."

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -248,42 +268,60 @@
-.. note:: The eigenvectors of matrices are not unique, so any eigenvector multiplied by a constant remains
-          a valid eigenvector. This function may compute different eigenvector representations on
-          different device types. Usually the difference is only in the sign of the eigenvector.
+.. note:: The eigenvectors of a matrix are not unique. Any eigenvector may be multiplied by
```
"Any eigenvector may be multiplied by any non zero number."

This statement is true for any vector in R^n or C^n. Maybe something like:

"A matrix's eigenvectors are not unique (e.g. multiplying an eigenvector by a non-zero number produces another set of eigenvectors for the matrix)."

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -248,42 +268,60 @@
-          a valid eigenvector. This function may compute different eigenvector representations on
-          different device types. Usually the difference is only in the sign of the eigenvector.
+.. note:: The eigenvectors of a matrix are not unique. Any eigenvector may be multiplied by
+          any non zero number.  As such, the returned eigenvectors are normalized to have norm
```
I wonder if this should be in the main documentation and not a note? 

Some more questions:
- "norm" w.r.t. which norm?
- what does "largest real component" mean?

Lezcano replied:
The eigenvectors live in R^n (resp. C^n), so this is computed with respect to the canonical norm in R^n (resp. C^n) \sum x_i^2 (resp. \sum z*_i z_i). I do not think there's any confusion here as R^n is, by definition, an finite-dimensional vector space with an inner-product, which induces a norm. The same happens for C^n but with an inner product + some extra structure.

Largest real component means that, for every eigenvector `v` with norm `1`, you can also choose `e^{i\phi}v`. So we choose (well, LAPACK does)  `\phi` such that it minimises `\norm{Re(e^{i\phi}v)}`.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -248,42 +268,60 @@
+.. note:: The eigenvectors of a matrix are not unique. Any eigenvector may be multiplied by
+          any non zero number.  As such, the returned eigenvectors are normalized to have norm
+          `1` and largest real component.
+          In particular, the eigenvectors are not continuous functions of the inputs.
```
This sentence does not seem related to the other facts in this note

Lezcano replied:
The lack of continuity follows from the non-uniqueness, but the reason is too technical to write it here. It's basically the same deal as what I described here https://github.com/pytorch/pytorch/pull/56265#discussion_r615764813

That's the idea, but formalising it requires either one line using heavy machinery (not useful) or more linear algebra than you can fit in a note..

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -248,42 +268,60 @@
 
-.. note:: See :func:`torch.linalg.eigvals` for a related function that computes only eigenvalues.
-          However, that function is not differentiable.
+.. warning:: This function assumes that :attr:`input` is diagonalizable (e.g. when all the
```
This assumption should be in the main documentation (see comment on "diagonalizable" above, too). 

If we also talk about input as if it's batched then this language will need a tweak, too.

Lezcano replied:
See the answer above.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -248,42 +268,60 @@
-          However, that function is not differentiable.
+.. warning:: This function assumes that :attr:`input` is diagonalizable (e.g. when all the
+             eigenvalues are different). If it is not, it will return the correct eigenvalues,
+             but the eigenvectors will be incorrect.
```
Can we better articulate how they're incorrect? "Incorrect" is so vague that maybe we should document the eigenvectors as "undefined" in this case?

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -248,42 +268,60 @@
 
-Keyword args:
-    out (tuple, optional): tuple of two tensors to write the output to. Default is `None`.
+        :func:`torch.linalg.eigvals` for a related function that computes only eigenvalues.
```
"for a related function that" -> ""

would be more consistent with prior templating; also the function is inherently related because it appears in this section

mruberry replied:
"only eigenvalues" -> "only the matrices' eigenvalues"

Lezcano replied:
Let's meet in the middle and write "computes only the eigenvalues"?

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -248,42 +268,60 @@
+        for Hermitian and symmetric matrices.
+
+        :func:`torch.linalg.svd` for a function that computes another type of spectral
+        decomposition. The SVD decomposition works on matrices that are not necessarily square.
```
what does "not necessarily square" mean?

Lezcano replied:
Changed to
>  :func:`torch.linalg.svd` for a function that computes another type of spectral decomposition that works on square and non-square matrices.

I wanted to emphasise the point that `eig` works on square matrices, but `svd` works also on non-square matrices.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -248,42 +268,60 @@
+        :func:`torch.linalg.svd` for a function that computes another type of spectral
+        decomposition. The SVD decomposition works on matrices that are not necessarily square.
+
+        :func:`torch.linalg.qr` for another (much faster) decomposition that works on general
```
what is a "general matrix"?

Lezcano replied:
Changed to:

>  :func:`torch.linalg.qr` for another (much faster) decomposition that works on square and non-square matrices.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -309,26 +347,25 @@
 
-Supports input of float, double, cfloat and cdouble dtypes.
+Supports :attr:`input` of float, double, cfloat and cdouble dtypes.
+The eigenvalues will always be complex-valued, even if :attr:`input` is real.
```
See previous notes about clarifying mathematical statements vs. statements about returned tensors and placement of sentences like this.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -309,26 +347,25 @@
+          CPU and CUDA inputs, respectively.
 
-.. note:: See :func:`torch.linalg.eig` for a related function that computes both eigenvalues and eigenvectors.
+.. note:: This function doesn't support backpropagation. Use :func:`torch.linalg.eig`
```
"doesn't support backpropagation" -> "is not differentiable."

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -309,26 +347,25 @@
 
-.. note:: See :func:`torch.linalg.eig` for a related function that computes both eigenvalues and eigenvectors.
+.. note:: This function doesn't support backpropagation. Use :func:`torch.linalg.eig`
+          instead, which also computes the eigenvectors.
```
Is linalg.eig actually differentiable now?

Lezcano replied:
I'll implement that this week.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -344,49 +381,78 @@
+
+    A = V \operatorname{diag}(L) V^{\text{H}}
+
+where :math:`V^{\text{H}}` denotes the conjugate transpose in the complex case and the transpose in
```
See previous comments for a suggestion for how to template this sentence

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -344,49 +381,78 @@
+    A = V \operatorname{diag}(L) V^{\text{H}}
+
+where :math:`V^{\text{H}}` denotes the conjugate transpose in the complex case and the transpose in
+the real case. The matris :math:`V` (and thus :math:`V^{\text{H}}`)
```
"matris" -> "matrix"

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -344,49 +381,78 @@
-Since the matrix or matrices in :attr:`input` are assumed to be Hermitian, the imaginary part of their diagonals
-is always treated as zero. When :attr:`UPLO` is "L", its default value, only the lower triangular part of each
-matrix is used in the computation. When :attr:`UPLO` is "U" only the upper triangular part of each matrix is used.
+:attr:`input` is assumed to be Hermitian (resp. symmetric), but this is not checked internally.
```
I feel like these facts are out of order. I was just reading about the return types and properties of the returned tensors and now I'm reading about input requirements.

Lezcano replied:
Moved up this paragraph.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -344,49 +381,78 @@
-is always treated as zero. When :attr:`UPLO` is "L", its default value, only the lower triangular part of each
-matrix is used in the computation. When :attr:`UPLO` is "U" only the upper triangular part of each matrix is used.
+:attr:`input` is assumed to be Hermitian (resp. symmetric), but this is not checked internally.
+When :attr:`UPLO` is `'L'` (default), only the lower triangular part of each matrix is used in the computation.
```
Maybe this can be documented in the arg descriptions?

Lezcano replied:
See https://github.com/pytorch/pytorch/pull/56265#discussion_r615922987

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -344,49 +381,78 @@
 
-Supports input of float, double, cfloat and cdouble dtypes.
+Supports :attr:`input` of float, double, cfloat and cdouble dtypes.
+The eigenvalues will always be real-valued, even if :attr:`input` is complex.
```
See previous comments

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -344,49 +381,78 @@
-.. note:: The eigenvalues/eigenvectors are computed using LAPACK's `syevd` and `heevd` routines for CPU inputs,
-          and MAGMA's `syevd` and `heevd` routines for CUDA inputs.
+.. note:: This function is computed using LAPACK's and MAGMA's `syevd` and `heevd` routines for
+          CPU and CUDA symmetric and Hermitian inputs respectively.
```
That "respectively" is doing a lot of work in this sentence.

"This uses LAPACK's \`syevd\` for symmetric CPU inputs and MAGMA's \`syevd\` for symmetric CUDA inputs. Hermitian CPU inputs use LAPACK's \`heevd\` and Hermitian CUDA inputs use MAGMA's \`heevd\`."

Lezcano replied:
That's much better. I did not know how to write this one, really.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -344,49 +381,78 @@
-.. note:: The eigenvectors of matrices are not unique, so any eigenvector multiplied by a constant remains
-          a valid eigenvector. This function may compute different eigenvector representations on
-          different device types. Usually the difference is only in the sign of the eigenvector.
+.. warning:: The eigenvectors of a matrix are not unique. Any eigenvector may be multiplied by `-1`
```
The second sentence here isn't actually saying anything because it just says that vectors can be multiplied by certain scalars.

Lezcano replied:
Fixed

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -344,49 +381,78 @@
+.. warning:: The eigenvectors of a matrix are not unique. Any eigenvector may be multiplied by `-1`
+             in the real case or by :math:`e^{i \phi}` for any :math:`\phi \in \mathbb{R}` in the
+             complex case.
+             As such, the eigenvectors are not continuous functions of the inputs, and different
```
Once again I'm not sure how the continuity of the function is relevant?

Lezcano replied:
See https://github.com/pytorch/pytorch/pull/56265#discussion_r615973559

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -344,49 +381,78 @@
+             in the real case or by :math:`e^{i \phi}` for any :math:`\phi \in \mathbb{R}` in the
+             complex case.
+             As such, the eigenvectors are not continuous functions of the inputs, and different
+             platforms, like NumPy, or inputs on different devices, may produce different
```
I feel like this is the heart of this warning, which should probably be downgraded to a note. Warnings are more appropriate for when something is likely to bite a user. Here this is just a note that different platforms might produce different eigenvectors because eigenvectors are not unique (and there's no canonical set of eigenvectors).

Lezcano replied:
I figured that the lack of continuity is something weird, and that most people could think that these maps are continuous, hence the warning.

I'd be happy to downgrade this to a note, if you think this is not too problematic, should we do that?

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -344,49 +381,78 @@
 
-.. note:: See :func:`torch.linalg.eigvalsh` for a related function that computes only eigenvalues.
-          However, that function is not differentiable.
+.. warning:: If `V` is used in the loss function, the gradient will only be finite when
```
What loss function?

mruberry replied:
Gradient of what?

Lezcano replied:
What I want to say is that, if `V.grad.is_defined()`, then...

How could I write that?

This happens also in the other spectral decompositions. I will update them once we come up with a good way of saying this..

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -344,49 +381,78 @@
 
-Keyword args:
-    out (tuple, optional): tuple of two tensors to write the output to. Default is ``None``.
+        :func:`torch.linalg.eigvalsh` for a related function that computes only eigenvalues.
```
saying the function is related is redundant with its appearing in this section

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -344,49 +381,78 @@
-Returns:
-    (Tensor, Tensor): A namedtuple (eigenvalues, eigenvectors) containing
+        :func:`torch.linalg.cholesky` for a different decomposition of a Hermitian matrix.
+        The Cholesky decomposition gives less information about the matrix but it is much faster
```
How comfortable are we making definitive statements about speed? Will a user complain if our cholesky decomposition is ever slower because we're using some goofball math library to compute it?

Lezcano replied:
I think it's great to put these, and if anything, we should try to make them true.

This is what the theory says, and PyTorch should offer similar guarantees. I also just put the very obvious partial order:
cholesky ~ qr < eigh < eig ~ svd.

If any of those strict comparisons do not hold, we should feel very bad and really try to fix it in my opinion.

This comment also goes in the line of my secret plan of teaching the user linear algebra while reading the docs, along the lines to the comments linking `inv` with `solve` and `pinv` with `lstsq`.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -344,49 +381,78 @@
+        to compute than the eigenvalue decomposition.
+
+        :func:`torch.linalg.eig` for a (slower) function that computes the eigenvalue decomposition
+        of a not necessarily Hermitian matrix.
```
what does "not necessarily Hermitian" mean?

Lezcano replied:
I want to emphasise that that function is slower but more general than this one, as this one just works for Hermitian matrices, and that one works for general matrices.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -344,49 +381,78 @@
+        of a not necessarily Hermitian matrix.
+
+        :func:`torch.linalg.svd` for a (slower) function that computes the more general SVD
+        decomposition of a not necessarily square matrix.
```
"not necessarily square matrix" -> "of a matrix with any shape" ?

Lezcano replied:
I really like this wording! I've changed all the occurences to use this wording.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -344,49 +381,78 @@
-            The eigenvalues in ascending order.
-        - **eigenvectors** (*Tensor*): Shape `(*, m, m)`.
-            The orthonormal eigenvectors of the :attr:`input`.
+        :func:`torch.linalg.qr` for another (much faster) decomposition that works on general
```
"general matrices" == ?

mruberry reviewed (approved):
Hey @Lezcano! This is a massive, necessary, and important PR. 

I got about halfway through, and my comments are increasingly of the form "see prior comment for how it applies here," so I'm going to stop for now. I'm very curious to hear your thoughts.

Overall I think we need to be very careful with language and organization of these docs, particularly around when we're describing the functions vs making mathematical statements. I also made some proposals for batched vs. unbatched templating. I think always treating the input tensor as batched (with possibly non-existent batch dimension) for operations that support batching will probably be the easiest thing and create the most consistent language, but I'm open-minded about other solutions to this issue.

Looking forward to an update!

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -12,55 +12,63 @@
 
-.. note:: When given inputs on a CUDA device, this function synchronizes that device with the CPU.
+.. note:: This function is computed using LAPACK's and MAGMA's `potrf` routine for
+          CPU and CUDA inputs, respectively.
```
Just to check, when you say batch size 1 you mean size `(n, n)`, not `(1, n, n)`, right?

xwang233 replied:
It applies to both `(n,n)` and `(1, n, n)` cases.

Lezcano replied:
Thank you!

@IvanYashchuk had a great idea on how to convey these things for the user: https://github.com/pytorch/pytorch/issues/54878#issuecomment-823002151

We'll probably clean and make these uniform after these PR, as it'd be good to go over all of them making sure that they are updated.

t10-13rocket reviewed (approved):
LGTM

lezcano committed:
Address comments

Lezcano commented:
Thanks for the great review @mruberry !

I addressed many of the comments, and left some answers for others.

I've also added in the first line of the documentation of `inv` and `eig` the note "if it exists". This should warn the user that there are some matrices that are not invertible (resp. diagonalisable) no matter what you do.

I've removed all the unicode in favour of LaTeX.

I have not updated the docs of `torch.svd` to match those of `torch.linalg.svd` yet. I will do that when we stabilise the docs for `torch.linalg.svd`.

lezcano committed:
Uniform batching template

lezcano committed:
Flaked

lezcano committed:
Flaked

lezcano committed:
Delete fortran-contiguous

IvanYashchuk commented on torch/linalg/__init__.py:
```diff
@@ -1438,49 +1642,57 @@
+  This behavior may change in a future PyTorch release.
 
-          * ``'reduced'`` : returns `(Q, R)` with dimensions (m, k), (k, n) (default)
+.. note:: The elements in the diagonal of `R` are not necessarily positive.
```
Is it in general expected that the diagonal elements are positive?

Lezcano replied:
I would say so. This is an easy way to make the QR unique, so it is often assumed in the literature. In other functions, such as `eig`, we have similar types of normalisation to make the decompositions unique when possible.

IvanYashchuk commented on torch/linalg/__init__.py:
```diff
@@ -1438,49 +1642,57 @@
+
+.. warning:: Backpropagation is only supported if the first `k = min(m, n)` columns
+             of every matrix in :attr:`input` are linearly independent.
+             If this condition is not met, no error will be thrown, but the gradient produced
+             may be anything.
```
"may be anything" is a bit unclear. Maybe "If this condition is not met, no error will be thrown, but the gradient produced is not correct"?

IvanYashchuk reviewed (commented):
@Lezcano, thank you for this work! It's really great to see this effort of unifying the style and language of the documentation. I left a few comments and I think it's good to be merged once Mike's feedback is addressed. 

I like the "seealso::" sections a lot, I think it would be useful for newcomers to navigate in the module.

lezcano committed:
Add and when talking about returned values
GPU -> CUDA

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -12,55 +12,62 @@
+    A = LL^{\text{H}}
+
+where :math:`L` is a lower triangular matrix and
+:math:`L^{\text{H}}` is the conjugate transpose when :math:`L` is complex, and the transpose when :math:`L` is real-valued.
```
I think it's good to be redundant, as many readers will not be used to work with complex numbers (cf. the issue that was opened a week ago about the use of `Vh` in the SVD).

Lezcano commented on torch/_torch_docs.py:
```diff
@@ -8329,38 +8334,34 @@ def merge_dicts(*dicts):
              and `the resulting vectors will span the same subspace`_.
              Different platforms, like NumPy, or inputs on different device types,
-             may produce different `U` and `V` tensors.
+             may produce different `U` and `Vh` tensors.
+
+.. warning:: The gradients with respect to `U` and `Vh` will only be finite when :attr:`input`
```
Yeah, sorry, these `torch.svd` docs are not updated, I should remove them from the PR. The `torch.linalg` version is updated though.

Lezcano commented on torch/_torch_docs.py:
```diff
@@ -8329,38 +8334,34 @@ def merge_dicts(*dicts):
-.. note:: The implementation of :func:`torch.linalg.svd` on CPU uses LAPACK's routine `?gesdd`
-          (a divide-and-conquer algorithm) instead of `?gesvd` for speed. Analogously,
+.. note:: The implementation of :func:`torch.linalg.svd` on CPU uses LAPACK's routine `gesdd`
+          (a divide-and-conquer algorithm) instead of `gesvd` for speed. Analogously,
```
Agree. Even then, we'll figure out a way to convey this information better with a table, as you suggested. As such, I'm leaving these things as they are for now, and then we can do another pass once we come up with a good template for these things.

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -12,55 +12,62 @@
 
     >>> a = torch.randn(3, 2, 2, dtype=torch.float64)
-    >>> a = torch.matmul(a, a.transpose(-2, -1))  # creates a symmetric positive-definite matrix
+    >>> a = a @ a.transpose(-2, -1) + torch.eye(2).squeeze(0)  # creates a batch of SPD matrices
```
But it's so loonnggg. Ok, changed :D

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -80,96 +87,96 @@
+    out (Tensor, optional): output tensor. Ignored if `None`. Default: `None`.
+
+Raises:
+    RuntimeError: if the :attr:`input` matrix or any matrix in the batch :attr:`input` is not invertible
```
Corrected all these mistakes in `RuntimeError`.

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -80,96 +87,96 @@
 
+.. seealso::
+
+        :func:`torch.linalg.slogdet` computes the sign and logarithm of the norm of
```
Corrected. Also made more explicit the real / complex case. This one is particularly annoying, because it's not completely obvious that the generalisation of sing / absolute value is the polar decomposition. Asking to know this would be too much for people without a background in complex numbers, so both cases have to be spelled out separately.

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -204,31 +211,36 @@
-The determinant can be computed with ``sign * exp(logabsdet)``.
+Computes the sign and logarithm of the norm of the determinant of a square matrix. Supports batched inputs.
+
+For a complex :attr:`input`, it returns the angle and the logarithm of the modulus of the
```
We can settle for one. I do not have any strong preference. What do you call |z| for a complex number? I like norm, but I'm happy with any of the options. I like "absolute value" a bit less, as it tends to be used a lot for real numbers.

Lezcano replied:
I changed it to "modulus" everywhere, is that fine?

IvanYashchuk replied:
Yes, I think it's fine since we are talking about math. Note that in programming it may mean a different thing: https://en.wiktionary.org/wiki/modulus#Noun

Lezcano replied:
Agree, but when you talk about the modulus in the other sense, it's always "the modulus with respect to  `p`" (or other number), so I think the confusion would be minimum. But thanks for bringing that up!

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -248,42 +260,65 @@
+
+.. math::
+
+    A = V \operatorname{diag}(L) V^{-1}
```
Agree that Lambda is a better notation, will change it. I prefer V for the matrix, as this is an invertible matrix (not necessarily orthogonal). Q is often used for orthogonal matrices and it'd be confusing.

I know that the V in SVD is orthogonal and the notation would clash, but we're starting to run out of letters in the latin alphabet...

I was thinking of using V for `eigh`, but perhaps I could use Q there as it is orthogonal?

IvanYashchuk replied:
Either choice for the eigenvectors is fine. No need to spend much time thinking about it 🙂

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -1214,34 +1343,41 @@
 
-Computes the pseudo-inverse (also known as the Moore-Penrose inverse) of a matrix :attr:`input`,
-or of each matrix in a batched :attr:`input`.
+Computes the Moore-Penrose pseudoinverse of a matrix. Supports batched inputs.
```
This is a tricky one. I've seen "Moore-Penrose inverse", "pseudoinverse" and "Moore-Penrose pseudoinverse" used.

What about writing:
>  Computes the pseudoinverse (Moore-Penrose inverse) of a matrix.

Then keep using pseudoinverse in the docs, as the operator is called `pinv`.

IvanYashchuk replied:
I think your suggestion is good.

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -80,96 +87,96 @@
 Args:
-    input (Tensor): the square `(n, n)` matrix or the batch of such matrices of size
-                    `(*, n, n)` where `*` is one or more batch dimensions.
+    input (Tensor): tensor of shape `(*, n, n)` where `*` is zero or more batch dimensions
+                    consisting of invertible matrices.
```
I agreed with a way to treat batched inputs in the docs of `torch.linalg`. I'll write it now, and it should solve this and other similar issues. Thanks for pointing it out.

lezcano committed:
Address points raised by Ivan

lezcano committed:
Changed input and other for A, B.
Make more explicit the size of the tensors in the definitions.
More uniform treatment of batching.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -92,11 +91,22 @@
-linalg.inv(input, *, out=None) -> Tensor
+linalg.inv(A, *, out=None) -> Tensor
+
+Computes the `inverse`_ of square matrix if it exists.
```
"of square matrix" -> "of a square matrix"

"if it exists" -> "if it exists. Throws a runtime error if the matrix is not invertible." ?

Lezcano replied:
About the second point. I think it's better to leave it as it is. That line is going to be the short summary it appears when you click on `torch.linalg` (similar to the one that we have in `torch.nn`). As such, I think the description should be "short and sweet". Then the user can enter each documentation to see what happens in each particular case.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -106,22 +116,22 @@
 
 Raises:
-    RuntimeError: if the :attr:`input` matrix or any matrix in the batch :attr:`input` is not invertible
+    RuntimeError: if the :attr:`A` matrix or any matrix in the batch :attr:`A` is not invertible.
```
"if the matrix :attr:\`A\` or any matrix in a batch of matrices :attr:\`A\` is not invertible."

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -92,11 +91,22 @@
+
+Letting :math:`\mathbb{K}` be :math:`\mathbb{R}` or :math:`\mathbb{C}`,
+for a matrix :math:`A \in \mathbb{K}^{n \times n}`,
+its inverse matrix :math:`A^{-1} \in \mathbb{K}^{n \times n}` is defined as
```
should this add a parenthetical to explain that the inverse may not exist?

"it inverse matrix :math:..., if it exists, is defined as" ?

mruberry replied:
Alternatively this could be reworded to explain that an inverse for a matrix is a matrix such that... and then it could say that if a matrix has an inverse it's "invertible" and the inverse is unique, but that not every matrix has an inverse, and those matrices are described as "not invertible."

What would you think about that approach?

Lezcano replied:
I'll do both. Put first the parenthesis (if it exists) and then describe in the next paragraph when may it not exist. Same for `torch.linalg.eig`.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -154,26 +164,28 @@
 
-.. note:: Backward through `det` internally uses :func:`torch.linalg.svd` when :attr:`input` is not
-          invertible. In this case, double backward through `det` will be unstable when :attr:`input`
+.. note:: Backward through `det` internally uses :func:`torch.linalg.svd` when :attr:`A` is not
```
Let's avoid use of the word "backward" (we're supporting forward AD soon!) and use phrases like "The gradient of..." So maybe in this case:

"The gradient of \`det\` is computed using ..."
"Because of this the gradient of the gradient of \`det\` is unstable when :attr:\`A\` doesn't have unique singular values..."

By "unstable" here, do we mean that the computation is incredibly numerically sensitive, or do we mean it's inaccurate?

Lezcano replied:
The "backward" point is a good one. I'll implement it throughout the docs.

About this note, I'm going to delete it, as it will become obsolete once Nik implements the new formula for the backward, which should be in soon.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -209,21 +221,20 @@
+Supports inputs of float, double, cfloat and cdouble dtypes.
+Also supports batched inputs, and, if the input is batched, the output is batched with the same dimensions.
 
+It returns a named tuple `(sign, logabsdet)`.
```
Should this be in a special "Returns" section? We're not consistent about having that section in the docs for every operator, but when an operator returns a named tuple that's typically where we elaborate on it

Lezcano replied:
This is a very good idea. The only one in which I am tempted to leave outside of `Return` is `lstsq`, as it currently has that crazy 4-tuple of a return. If put in the `Return` section, the list looks really crammed.

Lezcano replied:
In fact, I'm going to use that section to talk about the type of the returned tensor, whenever there is something to say about it (e.g. norm is always real even when the input tensor is complex).

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -258,44 +269,49 @@
 
-This decomposition exists if and only if :math:`A` is `diagonalizable`_. This is the case when all its eigenvalues are different.
+This decomposition exists if and only if :math:`A` is `diagonalizable`_.
+This is the case when all its eigenvalues are different.
```
I'd put this sentence immediately after the previous one. Suggested rewording:

"Equivalently, when the eigenvalues of :math:\`A\` are unique."

Lezcano replied:
The problem is that it is not a necessary condition, but only a sufficient one. The identity matrix is trivially diagonalisable but it has repeated eigenvalues.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -258,44 +269,49 @@
 
-Supports :attr:`input` of float, double, cfloat and cdouble dtypes.
-The output tensors `eigenvalues` and `eigenvectors` will always be complex-valued, even when :attr:`input` is real.
+The returned decomposition is a named tuple `(eigenvalues, eigenvectors)`,
```
Maybe a returns section for this tuple, too?

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -258,44 +269,49 @@
-          In particular, the eigenvectors are not continuous functions of the inputs.
+.. warning:: The eigenvectors of a matrix are not unique, as multiplying an eigenvector by a
+             non-zero number produces another set of valid eigenvectors of the matrix.
+             As such, the returned eigenvectors are normalized to have norm
```
I'm still not sure what "normalized to have norm \`1\` and largest real component" means

Previously we were converting the "norm" of a complex number to "absolute value" for a complex number,  would saying "absolute value" be more consistent here, too?

Is "scaled" a simpler term for "normalized"?

Lezcano replied:
I've rewritten this as you suggested, but I've also left the reason behind this problem in a new paragraph in the same warning:

>The eigenvectors of a matrix are not unique, nor are they continuous with respect to A. Due to this lack of uniqueness, different hardware and software may compute different eigenvectors.
This non-uniqueness is caused by the fact that multiplying an eigenvector by a non-zero number produces another set of valid eigenvectors of the matrix. In this implmentation, the returned eigenvectors are normalized to have norm 1 and largest real component.

Lezcano replied:
Same in `eigh` and `svd`.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -258,44 +269,49 @@
+             on different devices, may produce different singular vectors.
 
-.. warning:: This function assumes that :attr:`input` is `diagonalizable`_ (e.g. when all the
+.. warning:: This function assumes that :attr:`A` is `diagonalizable`_ (e.g. when all the
```
"e.g. when all the" -> "i.e. that its eigenvalues are unique"

Lezcano replied:
Will keep the wording for now, as per https://github.com/pytorch/pytorch/pull/56265#discussion_r619168317

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -345,14 +361,28 @@
-Computes the eigenvalues of a square matrix. Supports batched inputs.
+Letting :math:`\mathbb{K}` be :math:`\mathbb{R}` or :math:`\mathbb{C}`,
+the eigenvalues of a square matrix :math:`A \in \mathbb{K}^{n \times n}` are defined
+as the roots (counted with multiplicity) of the polynomial of degree `n`
```
"of a polynomial \`p\` of degree \`n\` such that"

Lezcano replied:
> The polynomial `p` of degree `n`  given by:

Corrected both occurrences

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -345,14 +361,28 @@
+
+The returned tensor will always be complex-valued, even when :attr:`A` is real.
+
+.. note:: The eigenvalues of a real matrix may be complex, as the roots of a real polynomial may be complex.
```
I would combine this note with the preceding sentence by extending it with:

"... even when :attr:\`A\` is real, because the roots of a real polynomial may be complex."

Lezcano replied:
The previous sentence has now been moved to the `Returns:`  section.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -379,33 +409,36 @@
-where :math:`V^{\text{H}}` is the conjugate transpose when :math:`V` is complex, and the transpose when :math:`V` is real-valued.
-The matrix :math:`V` (and thus :math:`V^{\text{H}}`) is orthogonal in the real case,
-and unitary in the complex case.
+:math:`Q^{\text{H}}` is the conjugate transpose when :math:`Q` is complex, and the transpose when :math:`Q` is real-valued.
```
"where..."

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -379,33 +409,36 @@
+- If :attr:`UPLO` `= 'L'` (default), only the lower triangular part of the matrix is used in the computation.
+- If :attr:`UPLO` `= 'U'`, only the upper triangular part of the matrix is used.
+
+The returned decomposition is represented as a namedtuple `(eigenvalues, eigenvectors)`,
```
Think about moving this to a "returns" section

Lezcano replied:
Done!

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -419,12 +452,12 @@
-             As such, the eigenvectors are not continuous functions of the inputs, and different
-             platforms, like NumPy, or inputs on different devices, may produce different
-             eigenvectors.
+             As this choice is not canonical, it happens that the eigenvectors are not continuous
```
See previous proposal for simplifying sections like this to avoid some tricky language.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -419,12 +452,12 @@
 
-.. warning:: If `V` is used to compute gradients, the gradient will only be finite when
-             :attr:`input` does not have repeated eigenvalues.
+.. warning:: If the `eigenvectors` are used to compute gradients, the gradient will only be
```
"Gradients computed using the \`eigenvectors\` tensor will only be finite when :attr:\`A\` has unique eigenvalues."

Lezcano replied:
Corrected here and in the `svd`. In my opinion the wording is a bit obscure, but I cannot suggest at the moment a better one.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -481,33 +514,44 @@
+- If :attr:`UPLO` `= 'L'` (default), only the lower triangular part of the matrix is used in the computation.
+- If :attr:`UPLO` `= 'U'`, only the upper triangular part of the matrix is used.
+
+The returned tensor will always be real-valued, even when :attr:`A` is complex.
```
I would try to combine these sentences and connect "tensor" with "eigenvalues":

"The returned eigenvalues will always be real-valued, even when :attr:\`A\` is complex, and sorted in ascending order."

Lezcano replied:
These sentences have been moved to the `Returns` section.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -633,38 +679,39 @@
+If :attr:`cond` `= None` (default), :attr:`cond` is set to the machine precision of the dtype of :attr:`A`.
 
-This function returns the solution to the problem and some extra information in a namedtuple of
+This function returns the solution to the problem and some extra information in a named tuple of
```
Consider moving this to a returns section

Lezcano replied:
This one is too big, and the return too complex. I'm leaving it here, but I'm also adding a `Returns` section with the names of the named tuple.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -633,38 +679,39 @@
-    The solution to this problem can be computed in terms of the Moore-Penrose pseudoinverse of
-    `A` as `A.pinv() @ B`.
-    For this reason, this function is a faster and more stable way of computing `A.pinv() @ B`.
+    A solution to this problem can be computed in terms of the pseudoinverse
```
This note reads a little oddly to me. The idea is that lstsq should be preferred to using the pseudo inverse, right? In that case I'd try:

"While solution = a.pinv() @ b, this function computes the solution in a faster and more numerically stable way that performing the computations separately."

Lezcano replied:
Corrected here and in `solve`.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -713,16 +760,19 @@
 If :attr:`n` `= 0`, it returns the identity matrix (or batch) of the same shape
-as :attr:`input`. If :attr:`n` is negative, it returns the inverse of each matrix
+as :attr:`A`. If :attr:`n` is negative, it returns the inverse of each matrix
 (if invertible) raised to the power of `abs(n)`.
```
What happens if the matrix is not invertible in this case?

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -713,16 +760,19 @@
+Supports inputs of float, double, cfloat and cdouble dtypes.
+Also supports batched inputs, and, if the input is batched, the output is batched with the same dimensions.
 
 If :attr:`n` `= 0`, it returns the identity matrix (or batch) of the same shape
```
This is mixing matrix and batch language but to be consistent with the new style I think it should just act as if "A" is a matrix?

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -830,35 +880,36 @@
-linalg.vector_norm(input, ord=None, dim=None, keepdim=False, *, dtype=None, out=None) -> Tensor
+linalg.vector_norm(A, ord=None, dim=None, keepdim=False, *, dtype=None, out=None) -> Tensor
+
+Computes the vector norm of a vector.
```
"a vector norm"? Since the function offers a choice of norms?

Lezcano replied:
Corrected here and in `norm`.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -830,35 +880,36 @@
+Also supports batched inputs, and, if the input is batched, the output is batched with the same dimensions.
 
-- If :attr:`dim`\ `= None`, the norm of the flattened :attr:`input` will be computed.
+- If :attr:`dim`\ `= None`, the norm of the flattened :attr:`A` will be computed.
```
"the norm of the flattened :attr:\`A\` will be computed" -> ":attr:\`A\` will be flattened before a norm is computed"

Lezcano replied:
Corrected here and in `norm`.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -960,17 +1013,21 @@
-linalg.norm(input, ord=None, dim=None, keepdim=False, *, out=None, dtype=None) -> Tensor
+linalg.norm(A, ord=None, dim=None, keepdim=False, *, out=None, dtype=None) -> Tensor
+
+Computes the vector norm of a vector or the matrix norm of a matrix.
```
"Computes a vector or matrix norm." ?

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -1084,41 +1138,42 @@
+Supports inputs of float, double, cfloat and cdouble dtypes.
+Also supports batched inputs, and, if the input is batched, the output is batched with the same dimensions.
 
 The returned decomposition is a named tuple `(U, S, Vh)`
```
Maybe a returns section here, too?

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -1142,17 +1197,17 @@
-             As such, `U` and `Vh` are not continuous functions of the inputs. Furthermore, different
-             platforms, like NumPy, or inputs on different devices, may produce different
-             `U` and `Vh` tensors.
+             As this choice is not canonical, it happens that `U` and `V`  are not continuous
```
see previous comments for a suggested simplification of this language

If we do want to keep language like this we might want to substitute "unique" for "canonical"

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -1212,19 +1267,24 @@
 
-The condition number of `A` measures the numerical stability of the linear system `AX = B`
+The condition number of :attr:`A` measures the numerical stability of the linear system `AX = B`
 in some specific matrix norm.
```
I'm not in love with the language "in some specific matrix norm." Is numerical stability often "in a norm"?

Lezcano replied:
The numerical stability of a system is computed with respect to a norm.

I'm changing the wording to
> ... of the linear system `AX = B` with respect to a matrix norm.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -1212,19 +1267,24 @@
+linalg.cond(A, p=None, *, out=None) -> Tensor
 
-Computes the condition number of a matrix with respect to a norm. Supports batched inputs.
+Computes the condition number of a matrix with respect to a norm.
```
can we be more precise about the norm? "a matrix norm"?

Lezcano replied:
Good catch. Changed.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -1245,7 +1305,7 @@
 For :attr:`p` is one of `('fro', 'nuc', inf, -inf, 1, -1)`, this function uses
 :func:`torch.linalg.norm` and :func:`torch.linalg.inv`.
-As such, in this case, the matrix (or every matrix in the batch) :attr:`input` should be square
+As such, in this case, the matrix (or every matrix in the batch) :attr:`A` should be square
```
"should" is a weird word for documentation

What happens if the matrices aren't square are aren't invertible?

Also, are we consistent about use of the word "invertible"?

Lezcano replied:
Yes, we are consistent. If it's not invertible we raise an error. This is documetned in the `Raises` section.

I'm changing "should be" to "has to be".

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -1256,10 +1316,9 @@
 
 In these cases, it is computed using :func:`torch.linalg.svd`. For these norms, the matrix
-(or every matrix in the batch) :attr:`input` may be rectangular
+(or every matrix in the batch) :attr:`A` may be rectangular
```
period missing at end of this sentence

mruberry replied:
"may be rectangular" for a matrix really means "may have any shape," right?

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -1341,19 +1400,22 @@
+Supports inputs of float, double, cfloat and cdouble dtypes.
+Also supports batched inputs, and, if the input is batched, the output is batched with the same dimensions.
+
+If :attr:`hermitian`\ `= True`, :attr:`A` should be a complex Hermitian or real symmetric
```
again, let's avoid the word "should" - we can say that A is assumed to be complex Hermitian or real symmetric, but this also needs to describe what that assumption entails and what will occur if the matrix is not actually Hermitian

Lezcano replied:
Changed to "is assumed to be". Nothing happens when it's not Hermitian. The next sentence explains this

>In this case, this function will just use the lower-triangular half of the matrix.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -1419,69 +1481,71 @@
 
-- If :attr:`input`, :attr:`other` have shapes `(*, n, n)`, `(*, n)`, for every matrix :math:`A`
-  in :attr:`input` and vector :math:`b` in :attr:`other`, this function returns a vector :math:`x` such that
+Computes the solution of a square system of linear equations with unique solution.
```
"with unique solution" -> "with a unique solution"

"solution of" -> "solution to"

mruberry replied:
What happens if there's not a unique solution? Is the behavior different if there's no solution vs many solutions?

Lezcano replied:
Changed.

There's one solution iff A is invertible. If it's not, it throws an error, as documnted in `Raises:`.

Lezcano replied:
Also, added the point that this system has unique solution iff A is invertible.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -1419,69 +1481,71 @@
-- If none of the above apply and :attr:`input`, :attr:`other` have shapes `(*, n, n)`, `(n,)` (resp. `(n, k)`),
-  the vector (resp. matrix) :attr:`other` is broadcasted to have shape `(*, n)` (resp. `(*, n, k)`).
-  This function then  returns the solution of the resulting batch of systems of linear equations.
+Denote by `*` zero or more batch dimensions.
```
I would move this to a "returns" section (like some other entries)

Instead of "Denote by..." use a let expression like "Letting \`*` be zero or more batch dimensions, the shape of the output is:"

Lezcano replied:
I left this here because it's this odd behaviour of being able to pass a vector or a matrix, and then the even weirder broadcasting behaviour. I think we should just accept matrices on both sides with the same batch dimensions, but numpy compat...

Denote changed.

About the organisation, I could split it as I did for SVD and QR (full vs reduced version), but I think it is not worth it. I think we should leave this here to document the odd casting behaviour. I would just move things to the `Returns` section when they are somewhat unimportant / they follow from the mathematical treatment. I would not want to move this, as it's just "an oddity" that we support, and it could bite people that are not aware of it.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -1419,69 +1481,71 @@
+  `(*, n, k)` (a batch of matrices or "multiple right-hand sides"), this function returns `X` of shape
+  `(*, n)` or `(*, n, k)` respectively.
+- Otherwise, if :attr:`A` has shape `(*, n, n)` and  :attr:`B` has shape `(n,)`  or `(n, k)`, :attr:`B`
+  is broadcasted to have shape `(*, n)` or `(*, n, k)` respectively and returns the solution of the resulting
```
the "returns" here refers to B, but it should refer to the function

this section needs a rethink to clarify these cases

Lezcano replied:
I changed it a bit, but yeah, happy to take suggestions on how to restructure this section.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -1419,69 +1481,71 @@
+  batch of systems of linear equations.
 
+.. note::
+    A solution to this problem can be computed in terms of the inverse
```
See previous note on a similar section for a suggestion on how to clarify this

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -1640,23 +1704,26 @@
+Supports inputs of float, double, cfloat and cdouble dtypes.
+Also supports batched inputs, and, if the input is batched, the output is batched with the same dimensions.
+
+The returned decomposition is a named tuple `(Q, R)`
```
Period at end of this sentence, since the return value is a tuple consider moving it to a "returns" section to elaborate on it

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -1640,23 +1704,26 @@
+The returned decomposition is a named tuple `(Q, R)`
 
-The parameter :attr:`mode` controls the shape of the output. If :attr:`input` has shape
+The parameter :attr:`mode` controls the shape of the output. If :attr:`A` has shape
```
Saying the mode parameter controls the shape is a little odd; yes different modes can cause different shapes, but the mathematical intuition behind mode is not that the computation produces different shapes, right?

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -1676,20 +1743,21 @@
 
-.. warning:: The QR decomposition is only locally unique when the input matrix has
-             independent columns. If this is not the case, different platforms, like NumPy,
+.. warning:: The QR decomposition is only unique up to sign of the diagonal of `R` when the
```
"up to sign" -> "up to the sign"

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -1676,20 +1743,21 @@
+             If this is not the case, different platforms, like NumPy,
              or inputs on different devices, may produce different valid decompositions.
 
 .. warning:: Backpropagation is only supported if the first `k = min(m, n)` columns
```
"Backpropagation is only supported if..." -> "Gradient computation using ??? is only supported if..."

mruberry reviewed (commented):
Hey @Lezcano! This is an impressive PR that does a good job refining the documentation and making it more consistent. I made notes inline, here's what I'd like to propose we do:

- let's take the weekend, then would you produce a final revision incorporating your thinking and considering the notes in the latest review
- then I'll do a quick check of the rendered docs to verify there's no egregious formatting issues, and we'll land this PR
- then we can switch to focusing on the existing linear algebra docs and how we should alias or deprecate operators like torch.svd in favor of their linalg counterparts

Does that sound good?

lezcano committed:
Incorporated all the suggestions of the second review.
Improved the writing and structure of some parts.

lezcano committed:
Flake8
Split into sections

Lezcano commented:
@mruberry Thanks for the review!
I added almost all of the suggestions. I left some comments about those I did not add.
I also made the split into the categories proposed in https://github.com/pytorch/pytorch/issues/54878#issuecomment-816636171

Your plan sounds great, let's land this one, as it's already in a pretty good state, and move on to the alias / deprecation of functions in the `torch.` namespace.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8374,38 +8379,34 @@ def merge_dicts(*dicts):
 
-.. note:: The implementation of :func:`torch.linalg.svd` on CPU uses LAPACK's routine `?gesdd`
-          (a divide-and-conquer algorithm) instead of `?gesvd` for speed. Analogously,
+.. note:: The implementation of :func:`torch.linalg.svd` on CPU uses LAPACK's routine `gesdd`
```
but this is torch.svd, not torch.linalg.svd

I thought we were not bothering with torch.svd changes for now?

Lezcano replied:
Sorry, I just revered the file.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -246,44 +273,71 @@
-linalg.eig(input, *, out=None) -> (Tensor, Tensor)
+linalg.eig(A, *, out=None) -> (Tensor, Tensor)
+
+Computes the eigenvalue decomposition of a square matrix if it exists.
```
What does this function do if the eigenvalue decomposition does not exist?

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -246,44 +273,71 @@
+             This non-uniqueness is caused by the fact that multiplying an eigenvector by a
+             non-zero number produces another set of valid eigenvectors of the matrix.
+             In this implmentation, the returned eigenvectors are normalized to have norm
+             `1` and largest real component.
```
I still don't know what this means -- but let's try to tweak it in a follow-up

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -304,31 +358,48 @@
 
-.. note:: See :func:`torch.linalg.eig` for a related function that computes both eigenvalues and eigenvectors.
+.. note:: This function is computed using LAPACK's and MAGMA's `geev` for CPU and CUDA inputs,
+          respectively.
```
accidental linebreaks here?

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -342,51 +413,91 @@
+The eigenvalues are returned in ascending order.
+
+.. note:: If the inputs are symmetric, this function is computed using LAPACK's and MAGMA's
+          `syevd` for CPU and CUDA inputs,
```
more accidental linebreaks?

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -998,78 +1163,123 @@
-             and `the resulting vectors will span the same subspace`_.
-             Different platforms, like NumPy, or inputs on different device types,
-             may produce different `U` and `Vh` tensors.
+.. warning:: The `U` and `V` are not unique, nor are they continuous with respect to :attr:`A`.
```
"the returned \`U\` and \`V\` tensors" ?

mruberry reviewed (approved):
Wow, impressive work @Lezcano. I made some very small comments, the most important of which is whether we want to change torch_docs.py in this PR.

This just needs a rebase, then ping me when you'd like it merged!

lezcano committed:
Revert torch.svd

lezcano committed:
minor

lezcano committed:
Add svdvals
Factorizations -> Decompositions
Example:: -> Examples::
Reword some bits

lezcano committed:
Small errata

lezcano committed:
Correct SVD

lezcano committed:
Shorter svd docs

lezcano committed:
Errata SVD

lezcano committed:
Remove lapack/magma backend

lezcano committed:
Minor edits to lstsq

lezcano committed:
flake8
